# Setting Up Search Engine Marketing Campaigns on a Large Scale 

Starting and maintaining SEM accoutns can, and usually is, a daunting task. In addition to hitting your business goals (revenue, conversions, clicks, etc.), you also need to have a systematic structure that is easy to understand and maintain.

What I will be laying out here is  briefly my thoughts on how to structure an account, but in much more detail, techniques that can help build a large account with numerous products, in a flexible way.

I will start by sharing what I think are the most important elements of any SEM account, and how I think they can best be mapped. This is not an explanation of what these elements are, as I'm sure you know, but it is to make explicit my assumptions and views, so it is clear why I am doing what I am doing. 

### Keywords >> Ads >> Landing Pages 

The main process of using search engines hasn't changed much: you go to the search engine's page, type in some words (or increasingly say them these days), get a few results, select one, and click to go to your page of choice. 

Based on this, the three most important elements that form the bulk of how an account is built are the following: 

1. Keywords: Representing the intention of the user, and what they want. 
2. Ads: Your promise to the user to satisfy their need, "I hear you, I have what you want, this is how I can help you, please visit this page."
3. Landing pages: This is where you satisfy the promise that you made to the user. 

If you properly map keywords to ads, and send users to the right landing pages, you have achieved the major part of setting up your campaigns. There are many other important things like ad extensions, bidding, and campaign settings for example. It's not that they are any less important, they just don't require that much time to create or set up.   
This is also congruent with the main elements that determine quality score; expected CTR, ad relevance, and landing page experience.  
To be practically specific, the biggest part of setting up your campaigns is creating two tables, one for the keywords, and the other for the ads. This is what we will be doing, and will end up producing those two tables for a fictional account. 

I will be using the programming language Python with a few functions that I have developed.  
I encourage you to follow along on the [interactive version of this article](https://change-me.com), where you can make modifications to the code and see how that changes the results. Most of the coding that I will share is quite simple, and you should be able to play with it, even if you have no prior programming knowledge. 


## Generating Keywords

Let me try and deconstruct the process of keyword research, and how we come up with our final list of keywords.  
You first go to a keyword tool. You enter a few keywords that are clearly relevant to the product/service you are promoting. Then the tool gives you a few hundred similar keywords. After that, you go through them one by one, selecting the ones that you believe would be relevant.  
Once you are done with selecting the keywords, you export, format, and finally upload them.  
You then repeat the process for every product/service that you have. 

Let's assume we have a used cars website for our example, and try to create the campaigns. 

I want to focus a bit on the process that goes through our minds while selecting our keywords from the keyword tool. Let's start with the product "Honda", so we would enter something like "buy honda", "honda price", "honda for sale", and get the tool's suggestions.  
While considering the relevant keywords, there are two conditions that a keyword needs to satisfy for it to qualify as relevant: 
1. **Product:** It has to contain the product, "honda" in our case (or any synonym)
2. **Word:** It has to contain a verb or word that indicates an intention that I'm trying to satisfy (buying a used Honda in this case)

Now, if you think of the possible verbs/words that can express that intention, you will find that they are not that many. Let's brainstorm some: 

buy, price, shop, used, second hand, auto loan, car loan, auto finance, car finance, cheap, cheapest, best price, lease, leasing, certified.  

I'm sure you can think of others but the idea is that although there are an infinite number of combinations of words or variations, the words themselves are not that many. Of course using phrase match and modified broad match is crucial in allowing users to come up with combinations that we didn't think of. The important thing here is that we are protected from irrelevant keywords by making sure that the proper verb is there, together with the product name. Needless to say, the proper negative match keywords need to also be added, as an additional layer of relevancy/protection.

With this, let's manually construct our first ad group, using just the first three words as a demonstration.  
The way I do it is to simply concatenate "honda" with each of the words, once before and once after: 

buy honda  
honda buy  
price honda  
honda price  
used honda  
honda used  

This way, and with phrase match, you can be confident that you cover the majority of keywords that could contain "honda" with any of those three words. "buy honda in <location>", "best site to buy honda", etc. would all be covered. The corner case, is when someone has a word between the word and product. "buy 2015 model honda" for example. This can be covered by modified broad match.  

Assuming we are happy with the word list that we produced, now it is trivial to do what we did with the first three across the board. "word honda" and "honda word" for all the words that we came up with, and using three match types; exact, phrase, and modified broad.  
I'm sure you can now see how trivial it is to do the same for "honda accord", "honda civic", "toyota", "toyota corolla" and all the cars that you have on your website.  

As a result our job now, in coming up with our keywords consists of two steps: 
1. Get the product list together with their corresponding URLs: this is trivial, and your colleague or client can easily send you a list of all the products that they sell, with their landing pages.
2. Come up with a list of verbs/words that signal the intention we are trying to satisfy.  

Step two here is the critical one. You can still use your favorite keyword tool to come up with those words. You can actually do the exercise for several car makes or models, and brainstorm, until you are comfortable that your list of verbs/words is comprehensive enough, then you are basically done. Now we start coding to get the final result.  
The challenge now is to do this for a large number of products, and put the keywords in the proper format, the table that contains the campaign name, ad group name, keyword, and match type so we can upload it.  

In `advertools` ([a Python package for online marketing](https://github.com/eliasdabbas/advertools/)) the `kw_generate` function is the one we will use for this task.  Here are the main arguments that the function takes:

* `products`: Self-explanatory. In our used cars example, that would be 'honda', 'honda accord', 'toyota', etc.
* `words`: The verbs/words that we finalized.
* `match_types`: A list of up to four possible match types; Exact, Phrase, Modified, or Broad.
* `order_matters`: True or False. Whether or not the order of the words in the keyword matters. If True, then you will have "honda price" as well as "price honda", otherwise, you will only have "honda price". 
* `max_len`: The number of words in a keyword that you want. If you specify 3, for example, the function will combine honda with one word, as well as with two words so you end up with keywords consisting of two and three words. "honda buy", "honda price", as well as "honda buy price", "honda price buy", and so on. This preempts the different combinations that will eventually come up, and makes your coverage better. It also exponentially adds to the number of keywords you are generating. More than three is possible, but it becomes way too complicated and cumbersome to be useful. With two and three words per keyword, I think you should be getting very good coverage. 
* `campaign_name`: Whatever you want to name your campaign. The default is 'SEM_Campaign'.

Let's see how it works with one product and two words:


In [1]:
import advertools as adv
adv.kw_generate(products=['honda'], 
                words=['buy', 'price'], 
                max_len=2, 
                match_types=['Exact', 'Phrase', 'Modified'],
                order_matters=False,
                campaign_name='SEM_Cars')

Campaign Ad Group        Keyword Criterion Type Labels
0  SEM_Cars    Honda      honda buy          Exact    Buy
1  SEM_Cars    Honda      honda buy         Phrase    Buy
2  SEM_Cars    Honda    +honda +buy          Broad    Buy
3  SEM_Cars    Honda    honda price          Exact  Price
4  SEM_Cars    Honda    honda price         Phrase  Price
5  SEM_Cars    Honda  +honda +price          Broad  Price

As you can see above, all you need to think about are the words, and then supplying the function parameters is very easy. With one line of code, you get the final table, ready for uploading and launching your keywords.  
Note that each product will get its own ad group, and ad group names are capitalized for better readability. Each keyword is repeated however many match types you specified. Also, as a free side-effect you get labels for your keywords.  
The labels are basically the words in each keyword (excluding the product).  
What this means is that now you can easily filter all the "price" keywords, and compare them with the "buy" keywords. These will include all products that contain them: 'toyota price', 'bmw price', and so on. So now you have another dimension of analysis that you can use to analyze performance.  
In the above example we ended up with six keywords in total. Let's run the same function by specifying `max_len` as three:

In [2]:
adv.kw_generate(products=['honda'], 
                words=['buy', 'price'], 
                max_len=3, 
                order_matters=False,
                campaign_name='SEM_Cars')

Campaign Ad Group             Keyword Criterion Type     Labels
0  SEM_Cars    Honda           honda buy          Exact        Buy
1  SEM_Cars    Honda           honda buy         Phrase        Buy
2  SEM_Cars    Honda         +honda +buy          Broad        Buy
3  SEM_Cars    Honda         honda price          Exact      Price
4  SEM_Cars    Honda         honda price         Phrase      Price
5  SEM_Cars    Honda       +honda +price          Broad      Price
6  SEM_Cars    Honda     honda buy price          Exact  Buy;Price
7  SEM_Cars    Honda     honda buy price         Phrase  Buy;Price
8  SEM_Cars    Honda  +honda +buy +price          Broad  Buy;Price

Now we have nine keywords, because we are combining words together. Let's see what happens if we do the same but specify `order_matters` as `True`: 

In [3]:
keywords = adv.kw_generate(products=['honda'], 
                           words=['buy', 'price'], 
                           max_len=3, 
                           order_matters=True,
                           campaign_name='SEM_Cars')
keywords.head().append(keywords.tail())

Campaign Ad Group             Keyword Criterion Type     Labels
0   SEM_Cars    Honda           honda buy          Exact        Buy
1   SEM_Cars    Honda           honda buy         Phrase        Buy
2   SEM_Cars    Honda         +honda +buy          Broad        Buy
3   SEM_Cars    Honda         honda price          Exact      Price
4   SEM_Cars    Honda         honda price         Phrase      Price
25  SEM_Cars    Honda     price honda buy         Phrase  Price;Buy
26  SEM_Cars    Honda  +price +honda +buy          Broad  Price;Buy
27  SEM_Cars    Honda     price buy honda          Exact  Price;Buy
28  SEM_Cars    Honda     price buy honda         Phrase  Price;Buy
29  SEM_Cars    Honda  +price +buy +honda          Broad  Price;Buy

We get thirty (I'm just showing the first and last five rows). Note also that the the labels where the keywords have multiple words also have more than one word in them.  
Now let's do the full account.  
We first create the variable `car_words` that includes the words we came up with.

In [4]:
car_words = [
    'buy',
    'price',
    'shop',
    'used',
    'second hand',
    'auto loan',
    'car loan',
    'auto finance',
    'car finance',
    'cheap',
    'cheapest',
    'best price',
    'lease',
    'leasing',
    'certified'
]

I got a random list of some popular car makes and models, which we will be using. 

In [5]:
import pandas as pd
cars = pd.read_csv('cars_urls.csv')
print('Rows:', cars.shape[0])
cars.head().append(cars.tail())

Rows: 70


Final URL               product
0                   www.mycarsite.com/Ford                  Ford
1              www.mycarsite.com/Chevrolet             Chevrolet
2             www.mycarsite.com/Volkswagen            Volkswagen
3                  www.mycarsite.com/Honda                 Honda
4                    www.mycarsite.com/GMC                   GMC
65  www.mycarsite.com/Ford_E150_Club_Wagon  Ford E150 Club Wagon
66     www.mycarsite.com/Jeep_Wrangler_4WD     Jeep Wrangler 4WD
67              www.mycarsite.com/Saab_900              Saab 900
68      www.mycarsite.com/Plymouth_Voyager      Plymouth Voyager
69       www.mycarsite.com/Audi_A4_quattro       Audi A4 quattro

Getting the product names together with their URLs is crucial, because later when we do the ads, we will be using the same names for ad groups, and this is how we make sure that they are consistent with each other. These are fake URLs that I constructed with the random set of car makes and models that I have. 

In [6]:
import advertools as adv

In [7]:
cars_kw_df2 = adv.kw_generate(cars['product'].values,  # the 'product' column of the 'cars' table
                              car_words, 
                              max_len=2,
                              campaign_name='SEM_Cars')

# Done. We now print to see a sample

print('Rows:', format(cars_kw_df2.shape[0], ','))
cars_kw_df2.sample(10)

Rows: 6,300


Campaign                  Ad Group                              Keyword  \
1473  SEM_Cars                      Audi                      Audi best price   
225   SEM_Cars                Volkswagen                       buy Volkswagen   
3385  SEM_Cars   Dodge Dakota Pickup 2Wd         used Dodge Dakota Pickup 2WD   
6181  SEM_Cars          Plymouth Voyager           auto loan Plymouth Voyager   
1069  SEM_Cars                   Pontiac                   best price Pontiac   
1167  SEM_Cars                    Subaru                     certified Subaru   
2419  SEM_Cars               Honda Civic               best price Honda Civic   
5065  SEM_Cars        Pontiac Grand Prix       Pontiac Grand Prix car finance   
3093  SEM_Cars  Chevrolet S10 Pickup 2Wd  Chevrolet S10 Pickup 2WD best price   
4004  SEM_Cars        Chevrolet Corvette      +Chevrolet +Corvette +certified   

     Criterion Type       Labels  
1473          Exact   Best Price  
225           Exact          Buy  
3385         Phrase         Used  
6181         Phrase    Auto Loan  
1069         Phrase   Best Price  
1167          Exact    Certified  
2419         Phrase   Best Price  
5065         Phrase  Car Finance  
3093          Exact   Best Price  
4004          Broad    Certified

In [8]:
cars_kw_df3 = adv.kw_generate(cars['product'].values, 
                              car_words, 
                              max_len=3,
                              campaign_name='SEM_Cars')


print('Rows:', format(cars_kw_df3.shape[0], ','))
cars_kw_df3.sample(10)

Rows: 138,600


Campaign              Ad Group  \
100747  SEM_Cars    Toyota 4Runner 4Wd   
129156  SEM_Cars  Ford E150 Club Wagon   
47431   SEM_Cars      Chevrolet Camaro   
33429   SEM_Cars                  Audi   
42494   SEM_Cars  Ford F150 Pickup 4Wd   
2613    SEM_Cars             Chevrolet   
116244  SEM_Cars       Ford Bronco 4Wd   
101049  SEM_Cars         Toyota Celica   
14481   SEM_Cars                Nissan   
11254   SEM_Cars            Mitsubishi   

                                            Keyword Criterion Type  \
100747            lease Toyota 4Runner 4WD car loan         Phrase   
129156  Ford E150 Club Wagon car finance best price          Exact   
47431                leasing lease Chevrolet Camaro         Phrase   
33429                       lease Audi auto finance          Exact   
42494         +shop +Ford +F150 +Pickup +4WD +cheap          Broad   
2613                      Chevrolet lease certified          Exact   
116244      car finance Ford Bronco 4WD second hand          Exact   
101049                    car finance Toyota Celica          Exact   
14481                            Nissan lease cheap          Exact   
11254                 auto finance price Mitsubishi         Phrase   

                         Labels  
100747           Lease;Car Loan  
129156   Car Finance;Best Price  
47431             Leasing;Lease  
33429        Lease;Auto Finance  
42494                Shop;Cheap  
2613            Lease;Certified  
116244  Car Finance;Second Hand  
101049              Car Finance  
14481               Lease;Cheap  
11254        Auto Finance;Price

And we are done. The above are two samples of fifteen random rows each. The first table has 6,300 rows (keywords), and the second has 138,600. The only difference is that I specified `max_len` as two in the first and three in the second. 

In [9]:
print('max_len=2')
(cars_kw_df2
 .groupby(['Ad Group', 'Criterion Type'])
 .agg({'Keyword': 'count'}).head(9))

max_len=2


Keyword
Ad Group        Criterion Type         
Audi            Broad                30
                Exact                30
                Phrase               30
Audi A4 Quattro Broad                30
                Exact                30
                Phrase               30
Buick           Broad                30
                Exact                30
                Phrase               30

In [10]:
print('max_len=3')
(cars_kw_df3
 .groupby(['Ad Group', 'Criterion Type'])
 .agg({'Keyword': 'count'}).head(9))

max_len=3


Keyword
Ad Group        Criterion Type         
Audi            Broad               660
                Exact               660
                Phrase              660
Audi A4 Quattro Broad               660
                Exact               660
                Phrase              660
Buick           Broad               660
                Exact               660
                Phrase              660

The summaries above show the number of keywords that we have for each match type under each ad group for the first three as a sample. Another important effect of this approach is consistency. All ad groups have the exact same set of keywords, with the only difference of having the product name substituted. This consistency should make it easier to compare ad groups.  
It's the same amount of work to generate six keywords as well as 6,300 in this case, or by changing a number, we get 138,600.  

When specifying `max_len=3` we get some keywords that are more specific, and convey a stronger intent to buy, like "second hand honda price" or "buy used honda". A portion of those additional keywords, however, is not really more relevant. "car loan car finance toyota" for example. It wouldn't hurt to have those keywords in your account, and it is easy to remove them, especially that now we have labels. So, you can filter out rows, where the label is "Car Loan;Car Finance" for instance. You can take a look at one ad group as a sample, and identify the cases where this doesn't make much sense, and apply to the whole table. 

Now, maybe you want to target more generic car terms, which have a high volume, but lower likelihood of converting. I'm referring to product names like car, auto, autos, etc.  
Again, we can use the same approach. The words have already been defined, now we just have to think of a few ways of saying "car". 

In [11]:
car_names = ['car', 'cars', 'auto', 'autos', 'automobile', 'vehicle']
generic_kw = adv.kw_generate(car_names, car_words, campaign_name='SEM_Cars')
generic_kw['Ad Group'] = 'Generic'  # put them all in one ad group

Done.

In [12]:
print('Rows:', format(generic_kw.shape[0], ','))
generic_kw.sample(15)

Rows: 11,880


Campaign Ad Group                              Keyword Criterion Type  \
7465   SEM_Cars  Generic                     cheap shop autos         Phrase   
6427   SEM_Cars  Generic                 autos cheap car loan         Phrase   
7682   SEM_Cars  Generic          +lease +autos +second +hand          Broad   
2702   SEM_Cars  Generic                    +buy +cars +price          Broad   
6570   SEM_Cars  Generic                  autos lease leasing          Exact   
1065   SEM_Cars  Generic                 second hand car used          Exact   
7722   SEM_Cars  Generic              lease second hand autos          Exact   
787    SEM_Cars  Generic                        buy cheap car         Phrase   
11594  SEM_Cars  Generic          +best +price +shop +vehicle          Broad   
1429   SEM_Cars  Generic              car finance car leasing         Phrase   
2931   SEM_Cars  Generic                shop car finance cars          Exact   
1969   SEM_Cars  Generic               certified cheapest car         Phrase   
9067   SEM_Cars  Generic            auto loan automobile shop         Phrase   
9037   SEM_Cars  Generic  second hand auto finance automobile         Phrase   
2346   SEM_Cars  Generic                  cars car loan cheap          Exact   

                         Labels  
7465                 Cheap;Shop  
6427             Cheap;Car Loan  
7682          Lease;Second Hand  
2702                  Buy;Price  
6570              Lease;Leasing  
1065           Second Hand;Used  
7722          Lease;Second Hand  
787                   Buy;Cheap  
11594           Best Price;Shop  
1429        Car Finance;Leasing  
2931           Shop;Car Finance  
1969         Certified;Cheapest  
9067             Auto Loan;Shop  
9037   Second Hand;Auto Finance  
2346             Car Loan;Cheap

Your website might have other ways of grouping cars. For example, you might have a section for SUVs, or maybe electric cars, hybrid cars, Japanese cars, and so on. 

Again, you simply have to come up with two lists; one for the products (variations of "SUV" for example), and one for the words, and you can do it in one step. 

You want to be more specific and generate keywords for bmw 320d, bmw 330d, bmw325xi, bmw 325i? It's easy. 

To demonstrate with another example, let's generate travel keywords for the top destinations in the world. The following code gets the top destinations from [Wikipedia](https://en.wikipedia.org/wiki/List_of_cities_by_international_visitors), and creates the variable `cities` taken from the third column. 

In [13]:
top_cities = pd.read_html('https://en.wikipedia.org/wiki/List_of_cities_by_international_visitors')[0]
top_cities.to_csv('top_cities.csv', index=False)

In [14]:
top_cities = pd.read_csv('top_cities.csv')
cities = top_cities['City'].tolist()
top_cities.head()

RankEuromonitor  RankMastercard       City         Country  \
0              1.0            11.0  Hong Kong       Hong Kong   
1              2.0             1.0    Bangkok        Thailand   
2              3.0             2.0     London  United Kingdom   
3              4.0             6.0  Singapore       Singapore   
4              5.0             NaN      Macau           Macau   

   Arrivals 2016Euromonitor  Arrivals 2016Mastercard  \
0                25695800.0                8370000.0   
1                23270600.0               21470000.0   
2                19842800.0               19880000.0   
3                17681800.0               12110000.0   
4                16299100.0                      NaN   

  Growthin arrivalsEuromonitor Income(billions $)Mastercard  
0                       −3.1 %                         6.84  
1                        9.5 %                        14.84  
2                        3.4 %                        19.76  
3                        6.1 %                        12.54  
4                        5.9 %                          NaN

In [15]:
print('Number of cities:', len(cities))
cities[:10]

Number of cities: 137


['Hong Kong',
 'Bangkok',
 'London',
 'Singapore',
 'Macau',
 'Dubai',
 'Paris',
 'New York City',
 'Shenzhen',
 'Kuala Lumpur']

Now that we have our `products` list ready, we just need to think of verbs that convey an intention to travel. Let's call them `travel_words`.

In [16]:
travel_words = ['tickets', 'trips', 'flights', 'vacation',
                'holiday', 'travel', 'airfare']

In [17]:
travel_kw_df = adv.kw_generate(cities, 
                               travel_words, 
                               campaign_name='SEM_Travel')

Done. 

In [18]:
print('Rows:', format(travel_kw_df.shape[0], ','))
travel_kw_df.sample(15)

Rows: 57,540


Campaign      Ad Group                       Keyword Criterion Type  \
16080  SEM_Travel  Johannesburg  Johannesburg holiday flights          Exact   
38669  SEM_Travel        Rhodes              +flights +Rhodes          Broad   
49139  SEM_Travel        Beirut      +airfare +travel +Beirut          Broad   
9826   SEM_Travel       Pattaya        Pattaya airfare travel         Phrase   
8084   SEM_Travel        Prague    +Prague +vacation +flights          Broad   
49412  SEM_Travel        Geneva      +flights +travel +Geneva          Broad   
15227  SEM_Travel        Madrid    +Madrid +vacation +holiday          Broad   
28802  SEM_Travel          Doha       +flights +Doha +tickets          Broad   
10502  SEM_Travel         Milan               +Milan +tickets          Broad   
9959   SEM_Travel       Pattaya     +vacation +trips +Pattaya          Broad   
48527  SEM_Travel         Xi'An        +trips +flights +Xi'an          Broad   
29784  SEM_Travel        Edirne        airfare Edirne tickets          Exact   
38806  SEM_Travel        Rhodes         Rhodes airfare travel         Phrase   
56716  SEM_Travel  Philadelphia           Philadelphia travel         Phrase   
36835  SEM_Travel          Jeju         vacation tickets Jeju         Phrase   

                 Labels  
16080   Holiday;Flights  
38669           Flights  
49139    Airfare;Travel  
9826     Airfare;Travel  
8084   Vacation;Flights  
49412    Flights;Travel  
15227  Vacation;Holiday  
28802   Flights;Tickets  
10502           Tickets  
9959     Vacation;Trips  
48527     Trips;Flights  
29784   Airfare;Tickets  
38806    Airfare;Travel  
56716            Travel  
36835  Vacation;Tickets

We generated 57,540 keywords for the destinations and the above table shows a random subset of the keywords.  
Let's create keywords for a recipes website. The following code extracts the national dish for each country from [Wikipedia]('https://en.wikipedia.org/wiki/National_dish').

In [19]:
import requests
from bs4 import BeautifulSoup

page = 'https://en.wikipedia.org/wiki/National_dish'
resp = requests.get(page)
soup = BeautifulSoup(resp.text, 'lxml')
dishes =  [x.text.strip() for x in soup.select('b + a')]
pd.DataFrame(dishes, columns=['dishlist']).to_csv('dishes.csv', index=False)

In [20]:
dishlist =  pd.read_csv('dishes.csv')['dishlist'].tolist()
print('Number of dishes:', len(dishlist))
dishlist[:10] + ['...'] # show the first ten

Number of dishes: 244


['dish',
 'Kabuli Palaw',
 'Tavë kosi',
 'Couscous',
 'Escudella',
 'Muamba de galinha',
 'Pepperpot',
 'Asado',
 'Khash',
 'Keshi yena',
 '...']

In [21]:
recipe_words = ['recipe', 'how to make', 'how to prepare',
                'make', 'cook', 'how to cook',]
recipes_kw_df = adv.kw_generate(dishlist, recipe_words,
                                max_len=2,
                                campaign_name='SEM_Recipes')

In [22]:
print(format(recipes_kw_df.shape[0], ','), 'keywords')
recipes_kw_df.sample(15)

8,784 keywords


Campaign       Ad Group                     Keyword Criterion Type  \
736   SEM_Recipes      Kuli Kuli       Kuli Kuli how to cook         Phrase   
4473  SEM_Recipes           Taco                   Taco make          Exact   
4257  SEM_Recipes      Garudhiya              Garudhiya make          Exact   
6168  SEM_Recipes        Cassava                Cassava cook          Exact   
7756  SEM_Recipes           Beef            Beef how to cook         Phrase   
3767  SEM_Recipes     Käsknöpfle  +how +to +make +Käsknöpfle          Broad   
4401  SEM_Recipes  Thieboudienne          Thieboudienne make          Exact   
1929  SEM_Recipes     Ikan Pepes      how to make Ikan Pepes          Exact   
951   SEM_Recipes       Feijoada        Feijoada how to cook          Exact   
5438  SEM_Recipes          Adobo              +Adobo +recipe          Broad   
3932  SEM_Recipes         Nshima   +Nshima +how +to +prepare          Broad   
825   SEM_Recipes     Ema Datshi      how to cook Ema datshi          Exact   
8196  SEM_Recipes          Chili        how to prepare Chili          Exact   
2022  SEM_Recipes         Pupusa       Pupusa how to prepare          Exact   
803   SEM_Recipes     Ema Datshi          +Ema +datshi +make          Broad   

              Labels  
736      How To Cook  
4473            Make  
4257            Make  
6168            Cook  
7756     How To Cook  
3767     How To Make  
4401            Make  
1929     How To Make  
951      How To Cook  
5438          Recipe  
3932  How To Prepare  
825      How To Cook  
8196  How To Prepare  
2022  How To Prepare  
803             Make

## You're Invited!

I'm sure you can see a pattern here. For each industry there can be a fairly standardized set of words that work with that industry. So why not prepare a set of words for different industries, and in different languages, and make it easy for people to start campaigns?  
I created a [sheet that is downloadable and editable](https://docs.google.com/spreadsheets/d/1yTUvZqKDH_tv6zQUtMPWadpd2_NicyrbENPfkf38P98), which you can explore and make any changes you think are useful, so everyone can benefit from everyone's ideas. 

In [23]:
%%html
<img src="starter_words.png" width=300>

### Limitations and Possible Issues: 

1. **Multiple match types in one ad group:** I never understood why this is a bad thing, and I usually put different match types in the same ad group, because it's more manageable and streamlined. If you prefer that approach, you can simply concatenate the ad group name with the match type, and end up with 'Toyota - Phrase' and 'Toyota - Exact'. This will create those new ad groups for you. You just need to make sure your change is consistent with the ads table.  
2. **All in one campaign:** You might also prefer to have more control over budgets and targeting, and may want to put each product in its own campaign (which I think is a good idea). This is very easy, as all you have to do is duplicate the ad group column, and this will create them as campaigns. Combined with the previous step, you can end up with a campaign for Toyota, and ad groups for each of the match types that you are using, if you prefer this kind of structure. Just make sure your changes are consistent with the ads table, if you want to make any of these two campaign structure changes.
3. **Synonyms and spelling variations:** There is no straightforward solution to this one. If you have a travel campaign for example, all of the following product names might be used to indicate the same destination: London, LHR, LGW, LDN, UK, England. Or maybe you have a real estate site: 2 bedroom, two bedroom, 2br, 2bed might all be used for the same prupose. In some cases like travel, it might be easier to come up with state, city, country, and airport codes to get synonyms, and in some cases you will have to be creative. 
4. **Brand and sub-brand issues:** In our example we used 'toyota camry' and 'ford mustang' for example. But why not just use 'camry' as a word, and add the keyword 'buy camry'? The problem is with model names like Dodge Charger for instance. If you have a keyowrd 'buy charger' and not 'buy dodge charger', your budget is going to be vanishing sooner than you want, and you will even be penalized for being irrelevant. A possible solution might be to do include the model without the make only for the names that are clearly referring to the car model and not something else. 
5. **Generic word products:** In some cases like movie titles, it's not enough to have the movie name on its own. One of the movies released a few years ago was titled "Sing", so the keywords 'watch sing' or 'download sing' won't mean much. So in these types of keywords you would need to add a qualifier like 'the film' or 'the movie', to make it clear what you are targeting. 
6. **Negatives:** This is not an issue actually, and it is very easy to generate, especially that we have a set of uniform and consistent ad groups. For used cars it would be "new", "2019", "2020", "brand new", etc. In some cases, you will probably need to add specific negatives at the ad group level. For example there is London, UK, and London, Ontario in Canada. You would need to add -canada as a negative keyword if you are targeting the UK London and vice versa.  
6. **Other issues:** There must be some issues that I haven't thought of, and I'm hoping that you share them if you disagree with the approach, or if you have suggestions as to what might go wrong. 

To summarize the discussion on keywords, since getting product names and their URLs is trivial in most cases, your job boils down to creating a few words to combine them with. I'm sure if asked to come up with twenty words for learning courses, or booking a hotel, or buying a certain product, that you can easily do it in a few minutes. The discussion is now about those 20 - 30 words that make sense for you, and then you are basically done.  
>  ### With this approach, you are making a transition from keyword research to keyword "manufacturing"!

Now let's manufacture some ads. 

## Creating Ads

I'd like to discuss two main techniques for creating ads on a large scale:


1. Constructing ads (bottom up approach): Nothing really different or new here, and can mostly be done with any spreadsheet software. There are some minor benefits though. 
2. Splitting text to get ads (top-down approach): This approach capitalizes on text that might be on landing pages, and provides a technique for creating ads from that text. 



### Constructing Ads (bottom-up approach)

This is a simple way of creating ads. Creating each of the slots one by one, and grouping them into one table.  
The function `ad_create` automates this by allowing you to provide the following parameters: 

- `template`: This is the text of the part ad slot you are creating. It could be a header, or a description line. Produt names will dynamically be inserted into their proper position that you specify. For example, you can write "Get the latest {}", and the product name will go in the empty space within the curly braces. 
- `replacements`: A list of the products that you are selling. As a result you will get "Buy Used Honda", "Buy Used Toyota", "Buy Used BMW", and so on. 
- `fallback`: In case your product name is too long and won't fit in the maximum number of characters allowed, this should replace the long product name with a suitable generic word that might make sense. 
- `max_len`: The maximum number of characters allowed for that slot of the ad. This makes sure that the text together with replacement (or fallback) don't exceed a certain number.

This can be done for any ad slot where you want to insert a product name, and the other can be written normally. For example, we will use this function for Headline 1 and Description 1. The others would be normal static text.  
Let's see how it words for one ad slot


In [24]:
(adv.ad_create(template='Buy The Best Used {}',  # replacements will go betweent the brackets
              replacements=cars['product'],
              max_len=30,                       # includes the replacements
              fallback='Cars') # the word to insert if the text is longer than `max_len`
 [:25])                        # get the first 25 (sample)

['Buy The Best Used Ford',
 'Buy The Best Used Chevrolet',
 'Buy The Best Used Volkswagen',
 'Buy The Best Used Honda',
 'Buy The Best Used Gmc',
 'Buy The Best Used Mitsubishi',
 'Buy The Best Used Toyota',
 'Buy The Best Used Nissan',
 'Buy The Best Used Dodge',
 'Buy The Best Used Hyundai',
 'Buy The Best Used Buick',
 'Buy The Best Used Pontiac',
 'Buy The Best Used Subaru',
 'Buy The Best Used Jeep',
 'Buy The Best Used Saab',
 'Buy The Best Used Plymouth',
 'Buy The Best Used Audi',
 'Buy The Best Used Isuzu',
 'Buy The Best Used Eagle',
 'Buy The Best Used Chrysler',
 'Buy The Best Used Cars',
 'Buy The Best Used Cars',
 'Buy The Best Used Ford Mustang',
 'Buy The Best Used Cars',
 'Buy The Best Used Cars']

Product names are inserted where we specified. In the cases where the text would be longer than `max_len` (30 in this example), the `fallback` word "Cars" was inserted.  
Now we do it to construct the whole ads.

In [25]:
pd.set_option('display.max_colwidth', 90)
ads_df = pd.DataFrame({
    'Campaign': 'SEM_Cars',
    'Ad Group': [x.title() for x in cars['product']],
    'Headline 1': adv.ad_create('Buy Used {}', cars['product'], 'Cars', 30),
    'Headline 2': 'Find Your Dream Car',
    'Headline 3': 'Enjoy Special Financing Deals',
    'Description 1': adv.ad_create('Select From the Largest Collection of Certified {}s', cars['product'], 'Car', 90),
    'Description 2': 'Book Your Free No-Questions-Asked Test Drive Today',
    'Final URL': cars['Final URL']
})
ads_df.to_csv('ads_df.csv', index=False)
ads_df.sample(20)

Campaign                   Ad Group                     Headline 1  \
1   SEM_Cars                  Chevrolet             Buy Used Chevrolet   
12  SEM_Cars                     Subaru                Buy Used Subaru   
32  SEM_Cars             Toyota Corolla        Buy Used Toyota Corolla   
64  SEM_Cars          Subaru Legacy Awd     Buy Used Subaru Legacy Awd   
43  SEM_Cars          Volkswagen Passat     Buy Used Volkswagen Passat   
67  SEM_Cars                   Saab 900              Buy Used Saab 900   
8   SEM_Cars                      Dodge                 Buy Used Dodge   
58  SEM_Cars            Ford Bronco 4Wd       Buy Used Ford Bronco 4wd   
28  SEM_Cars        Gmc Sierra 1500 4Wd   Buy Used Gmc Sierra 1500 4wd   
27  SEM_Cars        Gmc Sierra 1500 2Wd   Buy Used Gmc Sierra 1500 2wd   
45  SEM_Cars  Dodge Ram 1500 Pickup 2Wd                  Buy Used Cars   
53  SEM_Cars    Dodge Dakota Pickup 4Wd                  Buy Used Cars   
11  SEM_Cars                    Pontiac               Buy Used Pontiac   
49  SEM_Cars          Mitsubishi Mirage     Buy Used Mitsubishi Mirage   
34  SEM_Cars   Chevrolet S10 Pickup 2Wd                  Buy Used Cars   
18  SEM_Cars                      Eagle                 Buy Used Eagle   
59  SEM_Cars          Toyota Tacoma 4Wd     Buy Used Toyota Tacoma 4wd   
60  SEM_Cars                 Dodge B150            Buy Used Dodge B150   
9   SEM_Cars                    Hyundai               Buy Used Hyundai   
39  SEM_Cars       Ford F250 Pickup 2Wd  Buy Used Ford F250 Pickup 2wd   

             Headline 2                     Headline 3  \
1   Find Your Dream Car  Enjoy Special Financing Deals   
12  Find Your Dream Car  Enjoy Special Financing Deals   
32  Find Your Dream Car  Enjoy Special Financing Deals   
64  Find Your Dream Car  Enjoy Special Financing Deals   
43  Find Your Dream Car  Enjoy Special Financing Deals   
67  Find Your Dream Car  Enjoy Special Financing Deals   
8   Find Your Dream Car  Enjoy Special Financing Deals   
58  Find Your Dream Car  Enjoy Special Financing Deals   
28  Find Your Dream Car  Enjoy Special Financing Deals   
27  Find Your Dream Car  Enjoy Special Financing Deals   
45  Find Your Dream Car  Enjoy Special Financing Deals   
53  Find Your Dream Car  Enjoy Special Financing Deals   
11  Find Your Dream Car  Enjoy Special Financing Deals   
49  Find Your Dream Car  Enjoy Special Financing Deals   
34  Find Your Dream Car  Enjoy Special Financing Deals   
18  Find Your Dream Car  Enjoy Special Financing Deals   
59  Find Your Dream Car  Enjoy Special Financing Deals   
60  Find Your Dream Car  Enjoy Special Financing Deals   
9   Find Your Dream Car  Enjoy Special Financing Deals   
39  Find Your Dream Car  Enjoy Special Financing Deals   

                                                                 Description 1  \
1                   Select From The Largest Collection Of Certified Chevrolets   
12                     Select From The Largest Collection Of Certified Subarus   
32             Select From The Largest Collection Of Certified Toyota Corollas   
64          Select From The Largest Collection Of Certified Subaru Legacy Awds   
43          Select From The Largest Collection Of Certified Volkswagen Passats   
67                   Select From The Largest Collection Of Certified Saab 900s   
8                       Select From The Largest Collection Of Certified Dodges   
58            Select From The Largest Collection Of Certified Ford Bronco 4wds   
28        Select From The Largest Collection Of Certified Gmc Sierra 1500 4wds   
27        Select From The Largest Collection Of Certified Gmc Sierra 1500 2wds   
45  Select From The Largest Collection Of Certified Dodge Ram 1500 Pickup 2wds   
53    Select From The Largest Collection Of Certified Dodge Dakota Pickup 4wds   
11                    Select From The Largest Collection Of Certified Pontiacs   
49          Select From The Largest Collection Of Certified Mitsubishi Mirages   


In some ads the name of the cars was too long for the maximum number of characters allowed, so it was replaced with "Cars". We were able to put the car names in Description 1, as it allows for more space.  
It's easy to think of variations of the ads above. Changing text, different order, different call to action and so on.  
I won't spend time about copy writing, split testing and the different ways you can communicate and capture the audience. This is a totally different subject, and extremely important, but I'm mainly focusing here on techniques that enable us to build ads on a large scale. As with keywords, every set of ads can be generated with one function call, so feel free to experiment with different versions and approaches, based on your strategy and brand guidelines.  

The main idea of this article is the proper mapping of keywords to relevant ads, and landing pages. After generating tens of thousands of keywords and hundreds of ads, we need to make sure that the mapping is correct.  
We now have two tables, one for keywords `cars_kw_df3`, and one for the ads `ads_df`. 
"df" is shor for DataFrame. It's basically a table, and the name used in the popular data science languages. It's just a naming convention to know what your variables are, and which refers to which.  
Both DataFrames have the same campaign name, so that is covered. The more important thing to make sure we got right, is to make sure that all ad groups in the keywords table exist in the ads table, and vice versa. We also need to make sure that the overlap is complete, meaning there are no ad groups in one table but not the other. Python's `set` data structure is perfect for that.  
Creating a set from a list of items achieves two things. First, it removes the duplicates, giving us a list of all unique elements. Second it allows us to do operations almost exactly the same as the mathematical set operations; union, intersection, difference, etc.  
The two lines below do the same thing from opposite angles. Once it checks the difference between `cars_kw_df3`'s 'Ad Group' column  with the 'Ad Group' column of the `ads_df` DataFrame, and the second time it does the opposite.  
Getting a result of `set()` means the empty set, so the mapping seems to be working. 

In [26]:
set(cars_kw_df3['Ad Group']).difference(set(ads_df['Ad Group']))

set()

In [27]:
set(ads_df['Ad Group']).difference(set(cars_kw_df3['Ad Group']))

set()

Here we make sure that the URL's in `ads_df` correspond to the ad group names in the same DataFrame:

In [28]:
(ads_df['Final URL']       # take the column 'Final URL'
 .str.split('/').str[-1]   # split it by "/" and take the last element
 .str.replace('_', ' ')    # replace underscores with spaces
 .str.title()              # put it in title case
 .eq(ads_df['Ad Group'])   # check that it equals the 'Ad Group' column
 .all())                   # confirm that all are equal

True


### Creating Ads by Splitting Long Text

The other technique I would like to discuss is about utilizing the descriptive text you might have on landing pages to your advantage.  
Sometimes, you might have great descriptive text that might actually be used as an ad (or part of one).  
Let's say you have this text on your landing page (this is copied from a real page actually): 

> Used 2015 BMW 5 Series 535i xDrive Sedan AWD for sale with Sport Package, Leather Seats, Driver Assistance Package, Sunroof/Moonroof, Power Package, Navigation System, Technology Package, Aluminum Wheels, Heat Package, Premium Package, Climate Package, Luxury Package, Light Package, SE Package, SL Package, Bluetooth, Backup Camera, Comfort Package, Sound Package, M Sport Package'

The text clearly contains the most important details about our product. It's also great for relevancy, to include text from the landing page, and boosts your quality score a little. It makes your ads more transparent, as this is what the user will see after clicking on the ad.  
The issue is how to utilize this text, when you have hundreds of descriptions like this. The challenge is to split this text insto slots, each with a speified maximum number of characters, and making sure that they are meaningful words. You don't want to have an ad "that has wo | rds that are split like this".  
We will use the `ad_from_string` function to achieve this.  
The function takes the following parameters: 
- `s` (string): Any sentenct/phrase that you want to split. 
- `slots`: A list of numbers with the maximum allowed characters for each ad slot. The default uses Google's text ad default (30, 30, 30, 90, 90, 15, 15), but you can change it any way you want. 
- `sep`: The separator with which to split words. Typically, you don't need to change this as text will be split by white space. Sometimes, you might have words separated by underscores or dashes. 
- `capitalize`: True of False, defaults to False. If True, the final ad will have each of the words capitalized, otherwise, the capitalization will remain unchanged from what you provide.  

Let's see how it would work with our example text. 


In [29]:
sample_text = ('Used 2015 BMW 5 Series 535i xDrive Sedan AWD for sale with Sport Package,' 
              'Leather Seats, Driver Assistance Package, Sunroof/Moonroof, Power Package,' 
              'Navigation System, Technology Package, Aluminum Wheels, Heat Package, Premium'
              'Package, Climate Package, Luxury Package, Light Package, SE Package, SL Package, ' 
              'Bluetooth, Backup Camera, Comfort Package, Sound Package, M Sport Package')

sample_text_split =  adv.ad_from_string(sample_text)
sample_text_split

['Used 2015 BMW 5 Series 535i',
 'xDrive Sedan AWD for sale with',
 'Sport Package,Leather Seats,',
 'Driver Assistance Package, Sunroof/Moonroof, Power Package,Navigation System, Technology',
 'Package, Aluminum Wheels, Heat Package, PremiumPackage, Climate Package, Luxury Package,',
 'Light Package,',
 'SE Package, SL',
 'Package, Bluetooth, Backup Camera, Comfort Package, Sound Package, M Sport Package']

Let's see the lengths of the different slots that resulted:

In [30]:
[len(x) for x in sample_text_split]

[27, 30, 28, 88, 88, 14, 14, 82]

All the lengths are less than the specified lengths. You also get an additional slot at the end (length 82), and this is for the remaining text that didn't fit in our specified slots.  
You might be happy with this ad as is. You might want to use it differently. For example, you might want to have specific phrases in your headlines, and use the description lines for the details of the product. All you have to do is to run the same function with the same text, but specify the slots as (90, 90). 

In [31]:
sample_text_9090 =  adv.ad_from_string(sample_text, slots=(90, 90))
sample_text_9090

['Used 2015 BMW 5 Series 535i xDrive Sedan AWD for sale with Sport Package,Leather Seats,',
 'Driver Assistance Package, Sunroof/Moonroof, Power Package,Navigation System, Technology',
 'Package, Aluminum Wheels, Heat Package, PremiumPackage, Climate Package, Luxury Package, Light Package, SE Package, SL Package, Bluetooth, Backup Camera, Comfort Package, Sound Package, M Sport Package']

In [32]:
[len(x) for x in sample_text_9090]

[87, 88, 201]

Obviously this technique can be used on other ad formats, as not all Google Ads text ads have this format. You can also use them with Facebook, or Twitter ads.  
The main risk here as that you might have ads, with incomplete sentences, that might not make sense. As there is no standard text that is used, you will have to check and make some modifications. In real life, data doesn't usually come in a neat form. I scraped some sample car descriptions from the web, so we can take a look at how this might work out. These are random car descriptions so they won't work with our campaign because it is just a small sample for demonstration, but it works fine if you have descriptions for each of the URLs that we have.  
Here is a sample of those descriptions.

In [33]:
pd.set_option('display.max_colwidth', 500)
car_descriptions = pd.read_csv('car_descriptions.csv')
car_descriptions.sample(10)

description
25                                                                                                                                                                                                                                                                                      Used 2018 Jaguar F-PACE 30t Premium AWD for sale with Sunroof/Moonroof
9   Used 2015 BMW 5 Series 528i xDrive Sedan AWD for sale with Sport Package, Leather Seats, Driver Assistance Package, Sunroof/Moonroof, Power Package, Navigation System, Alloy Wheels, Heat Package, Premium Package, Climate Package, Luxury Package, Light Package, SE Package, Bluetooth, Backup Camera, Comfort Package, Sound Package, M Sport Package
29                                                                                                                                                                                     Used 2016 Jaguar F-TYPE R AWD for sale with Leather Seats, Sunroof/Moonroof, Navigation System, Bluetooth, Premium Wheels, Backup Camera, Remote Start, Premium Package
52                                                                                                                                                                                                                                            Used 2015 Toyota RAV4 LE AWD for sale with LE Package, Bluetooth, Backup Camera, Navigation System, Alloy Wheels
46                                                                                                                                                                                                                                                                                          Used 2017 Toyota Corolla LE for sale with Bluetooth, Backup Camera
38                                                                                                                                                                                                                                                                                           Used 2016 Honda CR-V LX AWD for sale with Steel Wheels, Bluetooth
41                                                                                                                                                                                                                                                                                                   Used 2012 Honda Civic Coupe Si for sale with Alloy Wheels
12                                                                                                                                                                                                                                                                                  Used 2015 BMW 5 Series 535i xDrive Sedan AWD for sale with Aluminum Wheels
16                                                                                                                                                                                                                                                          Used 2017 Jaguar F-PACE 35t Premium AWD for sale with Bluetooth, Sunroof/Moonroof, Aluminum Wheels
44                                                                                                                                                                                                                                                                      Used 2016 Honda CR-V EX AWD for sale with Bluetooth, Sunroof/Moonroof, Aluminum Wheels

And here is one function call to split sixty sample texts to the default Google text ads format, and generate the final table: 

In [34]:

ads_df_from_str = pd.DataFrame([adv.ad_from_string(x) for x in car_descriptions['description']], 
                               columns=['Headline 1', 'Headline 2', 'Headline 3', 'Description 1',
                                        'Description 2', 'Path 1', 'Path 2', 'remaining'])
ads_df_from_str.sample(15)

Headline 1                      Headline 2  \
49      Used 2016 Toyota Corolla S   Plus for sale with Bluetooth,   
22         Used 2015 Jaguar XF 3.0     Portfolio AWD for sale with   
16     Used 2017 Jaguar F-PACE 35t       Premium AWD for sale with   
3      Used 2015 BMW 5 Series 550i      Sedan RWD for sale with LE   
40   Used 2012 Honda Fit Sport for    sale with Bluetooth, Leather   
52    Used 2015 Toyota RAV4 LE AWD       for sale with LE Package,   
15  Used 2018 Jaguar E-PACE P250 S    AWD for sale with Bluetooth,   
8      Used 2015 BMW 5 Series 535i  xDrive Sedan AWD for sale with   
32   Used 2017 Honda Civic Touring    for sale with Leather Seats,   
53   Used 2015 Toyota RAV4 XLE for            sale with Bluetooth,   
56   Used 2013 Toyota RAV4 XLE AWD        for sale with Bluetooth,   
4      Used 2016 BMW 5 Series 528i  xDrive Sedan AWD for sale with   
6      Used 2015 BMW 5 Series 535i   Sedan RWD for sale with Sport   
23   Used 2018 Jaguar XE S AWD for        sale with Leather Seats,   
34   Used 2016 Honda CR-V EX-L AWD        for sale with Bluetooth,   

                        Headline 3  \
49                 Aluminum Wheels   
22                  Leather Seats,   
16    Bluetooth, Sunroof/Moonroof,   
3          Package, Sport Package,   
40             Seats, Alloy Wheels   
52       Bluetooth, Backup Camera,   
15         Backup Camera, Aluminum   
8            Leather Seats, Driver   
32      Sunroof/Moonroof, Adaptive   
53        Sunroof/Moonroof, Chrome   
56                  Backup Camera,   
4    Sport Package, Leather Seats,   
6   Package, Leather Seats, Driver   
23      Driver Assistance Package,   
34   Leather Seats, Backup Camera,   

                                                                                Description 1  \
49                                                                                              
22     Sunroof/Moonroof, Navigation System, Bluetooth, Premium Wheels, Backup Camera, Climate   
16                                                                            Aluminum Wheels   
3        Sunroof/Moonroof, Navigation System, Bluetooth, Backup Camera, Remote Start, M Sport   
40                                                                                              
52                                                            Navigation System, Alloy Wheels   
15                                         Wheels, Navigation System, Adaptive Cruise Control   
8          Assistance Package, Sunroof/Moonroof, Power Package, Navigation System, Technology   
32                  Cruise Control, Navigation System, Alloy Wheels, Bluetooth, Backup Camera   
53                                                                                     Wheels   
56                                                          Sunroof/Moonroof, Aluminum Wheels   
4        Driver Assistance Package, Sunroof/Moonroof, Power Package, Navigation System, Alloy   
6          Assistance Package, Sunroof/Moonroof, Power Package, Navigation System, Technology   
23  Sunroof/Moonroof, Navigation System, Adaptive Cruise Control, Bluetooth, Comfort Package,   
34                                                             Sunroof/Moonroof, Alloy Wheels   

                                                                                Description 2  \
49                                                                                              
22                                                                                    Package   
16                                                                                              
3                                                                    Package, Climate Package   
40                                                                                              
52                                                                                              
15                                                            

As you can see, in some cases the splitting works perfectly, and in some others there is a lot of blank space. The solution to these situations should not be difficult. Wherever we have an empty slot, we can fill it with one of the phrases that we used when constructing the ads with the bottom up approach.  
Let's summarize the lengths of the descriptions that we have and see how they are distributed.  
We need to see how many ads we have that fit into one slot, two slots, and so on.

In [35]:
from itertools import accumulate
lengths = pd.cut([len(x) for x in car_descriptions['description']], 
       bins=list(accumulate([0, 30, 30, 30, 90, 90, 15, 15, 500])))
lengths.describe().reset_index()

categories  counts     freqs
0     (0, 30]       0  0.000000
1    (30, 60]       6  0.100000
2    (60, 90]      16  0.266667
3   (90, 180]      25  0.416667
4  (180, 270]       5  0.083333
5  (270, 285]       0  0.000000
6  (285, 300]       1  0.016667
7  (300, 800]       7  0.116667

No description is less than thirty characters long, so we will focus on the other slots mainly.  
Combining the two techniques, we will check if an ad slot has text in it, in which case we will leave it as is, otherwise we will insert the corresponding slot from the previous exercise.  
The following code accomplishes this. So, we will only be filling the blanks when we need to.  
First let's do it for Description 1, just as an example to see how it works, then we construct the whole table

In [36]:
ads_df_from_str['Headline 1'] = [x if x else 'Used Cars For Sale' for x in ads_df_from_str['Headline 2']]
ads_df_from_str['Headline 2'] = [x if x else 'Find Your Dream Car' for x in ads_df_from_str['Headline 2']]
ads_df_from_str['Headline 3'] = [x if x else 'Enjoy Special Financing Deals' for x in ads_df_from_str['Headline 3']]
ads_df_from_str['Description 1'] = [x if x else 'Select From the Largest Collection of Certified Cars' 
                                    for x in ads_df_from_str['Description 1']]
ads_df_from_str['Description 2'] = [x if x else 'Book Your Free No-Questions-Asked Test Drive Today' 
                                    for x in ads_df_from_str['Description 2']]
ads_df_from_str['Path 1'] = ['Used Cars' for x in range(len(ads_df_from_str))]
ads_df_from_str['Path 2'] = ['For Sale' for x in range(len(ads_df_from_str))]
ads_df_from_str.sample(10)

Headline 1                      Headline 2  \
1   xDrive Sedan AWD for sale with  xDrive Sedan AWD for sale with   
33        for sale with Bluetooth,        for sale with Bluetooth,   
26      for sale with Alloy Wheels      for sale with Alloy Wheels   
30           Touring for sale with           Touring for sale with   
32    for sale with Leather Seats,    for sale with Leather Seats,   
27       Premium AWD for sale with       Premium AWD for sale with   
53            sale with Bluetooth,            sale with Bluetooth,   
15    AWD for sale with Bluetooth,    AWD for sale with Bluetooth,   
57       for sale with LE Package,       for sale with LE Package,   
42         sale with Steel Wheels,         sale with Steel Wheels,   

                       Headline 3  \
1   Sport Package, Leather Seats,   
33                 Backup Camera,   
26  Enjoy Special Financing Deals   
30      Bluetooth, Leather Seats,   
32     Sunroof/Moonroof, Adaptive   
27             Navigation System,   
53       Sunroof/Moonroof, Chrome   
15        Backup Camera, Aluminum   
57   Sunroof/Moonroof, Navigation   
42                      Bluetooth   

                                                                           Description 1  \
1   Driver Assistance Package, Sunroof/Moonroof, Power Package, Navigation System, Alloy   
33                               Sunroof/Moonroof, Adaptive Cruise Control, Alloy Wheels   
26                                  Select From the Largest Collection of Certified Cars   
30                      Backup Camera, Sunroof/Moonroof, Remote Start, Navigation System   
32             Cruise Control, Navigation System, Alloy Wheels, Bluetooth, Backup Camera   
27                                                                      Sunroof/Moonroof   
53                                                                                Wheels   
15                                    Wheels, Navigation System, Adaptive Cruise Control   
57             System, XLE Package, Alloy Wheels, Bluetooth, Backup Camera, Remote Start   
42                                  Select From the Largest Collection of Certified Cars   

                                                                         Description 2  \
1   Wheels, Technology Package, Heat Package, Premium Package, Climate Package, Luxury   
33                                  Book Your Free No-Questions-Asked Test Drive Today   
26                                  Book Your Free No-Questions-Asked Test Drive Today   
30                                  Book Your Free No-Questions-Asked Test Drive Today   
32                                  Book Your Free No-Questions-Asked Test Drive Today   
27                                  Book Your Free No-Questions-Asked Test Drive Today   
53                                  Book Your Free No-Questions-Asked Test Drive Today   
15                                  Book Your Free No-Questions-Asked Test Drive Today   
57                                  Book Your Free No-Questions-Asked Test Drive Today   
42                                  Book Your Free No-Questions-Asked Test Drive Today   

       Path 1    Path 2  \
1   Used Cars  For Sale   
33  Used Cars  For Sale   
26  Used Cars  For Sale   
30  Used Cars  For Sale   
32  Used Cars  For Sale   
27  Used Cars  For Sale   
53  Used Cars  For Sale   
15  Used Cars  For Sale   
57  Used Cars  For Sale   
42  Used Cars  For Sale   

                                                                                       remaining  
1   Package, Audio Package, SL Package, Bluetooth, Backup Camera, Comfort Package, Sound Package  
33                                                                                                
26                                                                                                
30                                                                                                
32                                              

As these are random descriptions and don't correspond directly to the keywords we created above, the ad group names would not make sense. When getting such descriptions, it's important to make sure that each one is mapped to the same product name you used for the keywords table, and this way you make sure that all your campaigns are complete with keywords and ads.  
Some ad slots have text in them, but very little. This depends on your judgement. You can keep them as is, or you can change the rule in the last code that we used to fill in the blanks. Instead of checking if a slot is empty, we can instead check if the length of the slot is less than a certain number of characters. 

## Summary

Let me summarize the process of creating the two main tables for keywords and ads. 

1. Get product names and their URLs (optionally get descriptions as well)

URL | Product | Description (optional)
----------------------|-------------|--
/make_model_1 | make model 1 | Description 1
/make_model_2 | make model 2 | Description 2
/make_model_3 | make model 3 | Description 3
/make_model_4 | make model 4 | Description 4
/make_model_5 | make model 5 | Description 5

2. Think of verbs/words that convey the intention you are targeting  
`words = ['buy', 'price', 'shop', etc...]`
3. Generate the keywords with one line of code:   
`adv.kw_generate(products, words)`
4. Create the ads: 

In [38]:
pd.DataFrame({
    'Campaign': 'The same campaign name used for keywords',
    'Ad Group': 'The list of products used for keywords, make sure to use str.title()',
    'Headline 1' :    adv.ad_create(template, products, fallback, max_len=30) or 'some static text',
    'Headline 2' :    adv.ad_create(template, products, fallback, max_len=30) or 'some static text',
    'Headline 3' :    adv.ad_create(template, products, fallback, max_len=30) or 'some static text',
    'Description 1' : adv.ad_create(template, products, fallback, max_len=90) or 'some static text',
    'Description 1' : adv.ad_create(template, products, fallback, max_len=90) or 'some static text',
    'Final URL': 'The URLs that correspond to the products'
})

5. Repeat the last step to come up with a few versions/variations.
6. (Optional) if you have descriptions, use `adv.ad_from_string` to split descriptions to sub-strings.
7. (Recommended) use the same technique to build sitelinks, and ad other ad extensions.
8. Set your campaign settings and bids, and Launch!